jax/interpreters/batching.py

```diff
@@ -33,11 +33,13 @@ def batch(fun: lu.WrappedFun, axis_name, axis_size, in_dims, out_dim_dests,
   # anlogue of `jvp` in ad.py
   fun, out_dims_thunk = batch_subtrace(fun)
   return _match_axes(batchfun(fun, axis_name, axis_size, in_dims),
-                     axis_size, out_dims_thunk, out_dim_dests)
+                     axis_size, in_dims, out_dims_thunk, out_dim_dests)
 
 @lu.transformation
 def batchfun(axis_name, axis_size, in_dims, *in_vals):
   # analogue of `jvpfun` in ad.py
+  if axis_size is None:
+    axis_size, = {x.shape[d] for x, d in zip(in_vals, in_dims) if d is not not_mapped}
```

Comment:
Can't we do this check once in `def batch` instead of both in `batchfun` and `_match_axes`?

Reply:
Unfortunately, `batch` doesn't have access to `in_vals`. We could factor out another `lu.transformation_with_aux` that we call from `batch`, and pass in the result to both `batchfun` and `_match_axes`, but that seemed a bit more complicated to me.

The fix I'd _really_ like is if we can assume all callers have an int `axis_size` available, and never pass in `None` (so these functions never have to compute the axis size from the args). I think `pxla.vtile` is the only caller which doesn't meet that requirement right now, but I figured it'd be easier just to make this work again now and then I could check with you as to whether `pxla.vtile` could easily be factored so as to have the axis size available.

WDYT? I can explain better if that's unclear.